src/transformers/modeling_encoder_decoder.py

```diff
@@ -18,35 +18,56 @@
 import logging
 import os
 
-from torch import nn
-
 from .modeling_auto import AutoModel, AutoModelWithLMHead
-
+from .modeling_utils import PreTrainedModel
 
 logger = logging.getLogger(__name__)
 
 
-class PreTrainedEncoderDecoder(nn.Module):
+class EncoderDecoderModel(PreTrainedModel):
     r"""
-        :class:`~transformers.PreTrainedEncoderDecoder` is a generic model class that will be
+        :class:`~transformers.EncoderDecoder` is a generic model class that will be
         instantiated as a transformer architecture with one of the base model
-        classes of the library as encoder and (optionally) another one as
+        classes of the library as encoder and another one as
         decoder when created with the `AutoModel.from_pretrained(pretrained_model_name_or_path)`
-        class method.
+        class method for the encoder and `AutoModelWithLMHead.from_pretrained(pretrained_model_name_or_path)` class method for the decoder.
     """
 
     def __init__(self, encoder, decoder):
-        super().__init__()
+        assert encoder is not None, "The encoder has to be defined"
+        assert decoder is not None, "The decoder has to be defined"
+
+        config = self._init_config(encoder.config, decoder.config)
+        config.is_encoder_decoder = True
+        super().__init__(config)
+
         self.encoder = encoder
+        assert (
+            self.encoder.get_output_embeddings() is None
+        ), "The encoder {} should not have a LM Head. Please use a model without LM Head"
         self.decoder = decoder
 
+    def _init_config(self, encoder_config, decoder_config):
+        # decoder config is used as default config (important for generation)
```

Comment:
The class EncoderDecoderModel needs a config to be able to instantiate a PretrainedModel to be able to use the `generate()` function. The question is which config we use here. For the moment I just use the given config of the decoder. I think we should maybe also add a couple of assert statements to check that the configs are compatible here, but no sure what to check exactly. What do you think @sshleifer @thomwolf @LysandreJik